[feat]: 대시보드, 컬럼, 초대하기 모달 & Modal API 구현

k-jipgeria opened the pull request:
## #️⃣ 연관된 이슈

- #29 

## 🔨작업 내용

> 이번 PR에서 작업한 내용을 간략히 설명해주세요(이미지 첨부 가능)

- 공통 모달 컴포넌트(GenericModal.tsx) 리펙토링
- 칼럼 모달 구현
- 대시보드 모달 구현
- 초대 모달 구현
- 각각의 모달에서 사용하는 API구현 
[모달 시연 영상.zip](https://github.com/user-attachments/files/18212427/default.zip)

## 💬 리뷰 요구사항(선택)

> 리뷰어가 특별히 봐주었으면 하는 부분이 있다면 작성해주세요
- Input, Button을 import해서 사용하고 있습니다. 해당 컴포넌트 관련해서 아직 부족한 부분이 많은 것 같애서 피드백 바랍니다.
- 각각의 모달에 대한 어느 피드백이든 환영입니다. 주관적인 것도 괜찮습니다.
- 참고 사항: 
  - 각각의 **모달의 로직 자체가 비슷해서** 파일은 많아도 한 모달만 이해하시면 나머지 모달은 금방 이해되실 겁니다.
  - 스타일은 **피그마랑 비슷하게만** 해 놓았습니다. 이후 진행 할 기능 리펙토링 끝나면 금방 스타일 완성할 수 있습니다!
  - 초대하기 모달에서 이메일 형식 검사 기능은 요구사항은 아니지만, 당연한 기능이라 빠른 시일 내에 **리펙토링** 하겠습니다
  - 현재 모달 컴포넌트에서 prop으로 받는 것들은 해당 모달이 사용될 페이지에서 관리하는 state입니다. 
  페이지 파일에서 state로 모달과 연결해주어야 모달 확인이 가능합니다. 
  - 바로 로컬에서 시연해보고 싶은 분께서는 브랜치에서 test/modal 를 로컬에 당겨와서 시연해보시면 됩니다.

JIPGERIA committed:
refactor: 피그마와 모달 스타일이 비슷하도록 수정

JIPGERIA committed:
refactor: disabled 상태에서의 스타일 추가

JIPGERIA committed:
feat: 대시보드, 컬럼, 초대하기, 할 일 모달에 사용하는 input component

JIPGERIA committed:
feat(#29): 대시보드, 컬럼, 초대하기 모달 & API 구현

- 대시보드(CreateDashboardModal.tsx): 새로운 대시보드를 생성할 수 있는 모달입니다.
- 컬럼(CreateColumnModal.tsx, EditColumnModal.tsx): 각각 컬럼을 생성, 수정할 수 있는 모달입니다.
- 초대하기(InviteModal.tsx): 이메일을 입력하여 특정 대시보드 내에 Member를 추가할 수 있는 모달입니다.
- API(modalApi.ts): 위에 설명한 모달들의 API 로직이 모여있는 파일입니다.

- 참고 사항: error.ts는 임시로 만든 것임을 참고바랍니다.

Donghunn-Lee commented on app/_components/Modals/Column/EditColumnModal.tsx:
```diff
@@ -0,0 +1,114 @@
+  initialTitle?: string;
+}
+
+const EditColumnModal = ({ onClose }: EditColumnModalProps) => {
+  const initialTitle = 'initialTitle'; // 제거하시고 불러온 초기 컬럼 이름 사용하시면 됩니다
+  const columnId = 'columnId'; // 제거하시고 불러온 컬럼 ID 사용하시면 됩니다
+
```
불러온 값들을 사용하기 위해선 `prop`으로 선택한 column 정보를 받을 수 있어야 할 것 같습니다!

k-jipgeria replied:
저런 부분은 모달을 import해서 사용할 지, 모달을 페이지 파일 안에서 작성할 지에 따라 달라져서, 팀미팅 때 이야기 해 보시죠

yunrap replied:
페이지 파일안에서 모달을 정의하는방법이 더나아보입니다!~
공통컴포넌트GenericModal을 각페이지에 넣어주는게 더 통일성있을거같습니다

Donghunn-Lee commented on app/_components/Modals/Column/EditColumnModal.tsx:
```diff
@@ -0,0 +1,114 @@
+
+  // 컬럼 삭제
+  const handleDeleteColumn = async () => {
+    const isConfirmed = window.confirm(
+      '칼럼의 모든 카드가 삭제됩니다. 계속하시겠습니까?',
+    );
```
가능하면 위의 `alert`와 `confirm`은 브라우저 기본 기능 말고 모달창이었으면 더 좋을 것 같습니다. 나중에 생각해보시면 될 것 같아요!

k-jipgeria replied:
네 알겠습니다!

Donghunn-Lee reviewed (approved):
잘 작성된 것 같습니다. 수고 많으셨습니다!

bhoh032019 reviewed (commented):
createDashboardModal을 실행해 보니 

![image](https://github.com/user-attachments/assets/1033519c-7fbe-4b31-ba7f-6fce33caec78)


다음과 같은 에러가 발생합니다. 대시보드 생성은 되는데.. 제가 이용을 잘못한 걸까요??

yunrap commented on app/_components/Button/Button.tsx:
```diff
@@ -2,23 +2,38 @@ interface ButtonProps extends React.ButtonHTMLAttributes<HTMLButtonElement> {
+        ${customStyle}
+      `}
+      disabled={disabled}
+      {...props}
```
disabled 부분 필요했는데 추가 감사합니다!~

yunrap commented on app/_components/Modals/Dashboard/CreateDashboardModal.tsx:
```diff
@@ -0,0 +1,128 @@
+        customStyle="w-full h-10"
+        disabled={isLoading}
+      />
+
+      <div className="flex gap-3">
+        {COLOR_OPTIONS.map((color) => (
+          <button
+            key={color.label}
+            type="button"
+            className="focus:ring-primary flex h-10 w-10 items-center justify-center rounded-full transition-transform hover:scale-110 focus:outline-none focus:ring-2 focus:ring-offset-2"
+            onClick={() => setSelectedColor(color.value)}
+            style={{ backgroundColor: color.value }}
+            aria-label={`${color.label} 색상 선택`}
+            disabled={isLoading}
+          >
+            {selectedColor === color.value && (
+              <Image src={CheckIcon} alt="선택된 색상" width={20} height={20} />
+            )}
+          </button>
+        ))}
+      </div>
+    </div>
+  );
```
저랑 컴포넌트가 겹치네요!~ ㅎㅎ 
ColorSelectBox 컴포넌트 만들었습니다!~ 머지된후에 나중에 리팩토링에 추가하셔도 좋을거같아요!~

k-jipgeria replied:
맞습니다
리펙토링 때 고려해보겠습니다!

k-jipgeria replied:
네 경로 수정하겠습니다!

neruu00 commented on app/_components/Input/Input.tsx:
컴포넌트 이름이 단순해 어디서 사용하는지 파악하기 힘들 거 같아요!

k-jipgeria replied:
네 수정하겠습니다

neruu00 commented on app/_components/Button/Button.tsx:
개인적인 의견이지만 `customStyle` 속성은 어차피 `className`과 동일한 기능을 하는 거 같아 `className`으로 사용거나 이미 버튼의 속성을 상속했기 때문에 `customStyle`을 빼버려도 무방할 거 같습니다.

저희 팀 모두는 스타일링을 tailwindCSS로 하고 있기 때문에, `className`만 있어도 그 의도를 충분히 유추할 수 있을 거 같습니다!

yunrap commented on app/api/modalApi.ts:
```diff
@@ -0,0 +1,99 @@
+  Authorization: `Bearer ${ACCESS_TOKEN}`,
+});
+
+export class ModalApi {
+  static async createDashboard(
+    newDashboardData: CreateDashboardRequest,
+  ): Promise<DashboardResponse> {
+    const { title, color } = newDashboardData;
+
+    try {
+      const response = await instance.post<CreateDashboardResponse>(
+        '/dashboards',
+        { title, color },
+        { headers: createAuthHeader() },
+      );
+      return response.data.data;
+    } catch (error) {
+      throw handleApiError(error);
+    }
+  }
+
+  static async createColumn(
+    columnData: CreateColumnRequest,
+  ): Promise<ColumnResponse> {
+    const { title, dashboardId } = columnData;
```
이렇게 static으로 메서드를 관리할수있군요!!
전 useasync 함수많이 편하게사용해서 그런방법으로 형식을 통일시켜도 좋을거같아요!~

k-jipgeria replied:
알겠습니다!

neruu00 commented on app/_components/Input/Input.tsx:
이 부분의 `customStyle`도 위에서 언급한 것과 동일합니다!

k-jipgeria replied:
네 둘 중 하나만 남기겠습니다

neruu00 commented on app/_components/Modals/Column/CreateColumnModal.tsx:
```diff
@@ -0,0 +1,92 @@
+  dashboardId,
+}: {
+  onClose: () => Promise<void> | void;
+  dashboardId: string;
```
`dashboardId`는 `number` 타입으로 알고 있는데 이 부분은 한 번 더 확인 바랍니다!

k-jipgeria replied:
수정하겠습니다!

yunrap reviewed (changes requested):
모달을 import해서 사용할 지, 모달을 페이지 파일 안에서 작성할지 회의된후 다시 pr 리뷰달게요!~ㅎㅎ

neruu00 commented on app/_components/Modals/Column/EditColumnModal.tsx:
해당 모달 하나의 페이지에서만 사용하는 의존적인 컴폰넌트처럼 보이는데 `_components` 대신 `/dashboard/{dashboardid}` 경로 내부에 있어도 될 거 같습니다

neruu00 commented on app/_components/Modals/Dashboard/CreateDashboardModal.tsx:
해당 모달도 하나의 페이지에서만 사용하는 의존성 컴포넌트로 경로를 수정하는게 더 좋아보입니다!

k-jipgeria replied:
네!
의존성 컴포넌트의 의미를 살릴 수 있도록 경로 수정하겠습니다

neruu00 commented on app/api/modalApi.ts:
테스트로 만드신거면 상관없지면, 만약 테스트가 아니라면 modalAPI라는 이름은 역할을 파악하기 모호합니다

여러 부분으로 분할 해도 될 거 같은 코드를 하나의 클래스로 묶은 의도가 있을까요?

k-jipgeria replied:
모달 별로 분리하는 게 좋겠네요! 
리펙토링 때 추가하겠습니다

neruu00 commented on app/utils/error.ts:
```diff
@@ -0,0 +1,34 @@
+  }
+
+  throw new ApiError('알 수 없는 오류가 발생했습니다.', 500);
+};
```
`handle`의 경우 발생한 이벤트에 대한 처리를 할 경우 사용합니다. 에러도 하나의 이벤트로 볼 수 있지만 해당 함수는 에러에 대한 핸들링이라기 보단 에러를 ApiError라는 형태로 변형해서 throw하는 역할입니다.  

즉, handle보단 convert 혹은 throw에 더 적합한 거 같습니다.

k-jipgeria replied:
수정하겠습니다!

neruu00 commented on app/api/modalApi.ts:
```diff
@@ -0,0 +1,99 @@
+        { email },
+        { headers: createAuthHeader() },
+      );
+      return response.data.data;
```
해당 스크립트 내에서 `response.data.data`를 사용해서 AxiosResponse에서 데이터를 추출하는데 제가 알기로는 `response.data`만 사용해도 되는 것으로 아는데 이 부분 한 번 확인 바랍니다!

https://axios-http.com/docs/res_schema

neruu00 reviewed (approved):
고생하셨습니다!

아래 내용은 반드시 지켜주셨으면 좋겠습니다.

코드가 너무 길어 하나의 기능을 파악하는데 너무나 오랜 시간을 할애해야 합니다.

추가적으로 현재 저희는 squash merge를 사용하는데 해당 분기에서 심각한 오류가 발생해도 오류 부분만 덜어내기도 쉽지 않습니다.

하나의 PR에는 너무 많은 기능을 담지 않아 주셨으면 좋겠습니다.

만약 너무 변경사항이 많다면 새로운 분기를 만든 다음 의존적인 코드 단위로 나눠서 따로 PR을 올리는 방식을 부탁드립니다.

ghost commented:
코멘트 달아주신 것들 참고해서 리펙토링하겠습니다

JIPGERIA committed:
fix: response 제대로 사용할 수 있도록 return문 & type 수정

Donghunn-Lee commented:
![image](https://github.com/user-attachments/assets/cbe5302d-0393-4e87-bc4f-f1c72db2c945)

빌드 에러 발생중입니다 확인 부탁드려요!

JIPGERIA committed:
fix: dashboardId type error 수정

ghost commented:
> 고생하셨습니다!
> 
> 아래 내용은 반드시 지켜주셨으면 좋겠습니다.
> 
> 코드가 너무 길어 하나의 기능을 파악하는데 너무나 오랜 시간을 할애해야 합니다.
> 
> 추가적으로 현재 저희는 squash merge를 사용하는데 해당 분기에서 심각한 오류가 발생해도 오류 부분만 덜어내기도 쉽지 않습니다.
> 
> 하나의 PR에는 너무 많은 기능을 담지 않아 주셨으면 좋겠습니다.
> 
> 만약 너무 변경사항이 많다면 새로운 분기를 만든 다음 의존적인 코드 단위로 나눠서 따로 PR을 올리는 방식을 부탁드립니다.

네 다음부터 
- 에러 발생 시 에러만 제거하기 어려움이 있으므로 기능 별로 나누어서 PR하겠습니다
- 여러 기능이 있을 시 하위 브랜치에 나누어서 PR하겠습니다

ghost commented:
> createDashboardModal을 실행해 보니
> 
> ![image](https://private-user-images.githubusercontent.com/162112399/397902014-1033519c-7fbe-4b31-ba7f-6fce33caec78.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MzQ3ODk2NjIsIm5iZiI6MTczNDc4OTM2MiwicGF0aCI6Ii8xNjIxMTIzOTkvMzk3OTAyMDE0LTEwMzM1MTljLTdmYmUtNGIzMS1iYTdmLTZmY2UzM2NhZWM3OC5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQxMjIxJTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MTIyMVQxMzU2MDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT1jYmQzZDc3MzdhNGE0OTIwMjc2ZTMxMGRiYTVkNzM2OGQ2MTNlZWUyMTExODkyOGM4YjcxYWIyZWVkMjkyN2YwJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCJ9.55E1mzs763WNNnvVTninX62VihTcEhg3Npilt32NHJQ)
> 
> 다음과 같은 에러가 발생합니다. 대시보드 생성은 되는데.. 제가 이용을 잘못한 걸까요??

수정했습니다!

JIPGERIA committed:
Merge branch 'dev' into feat/29--modal

JIPGERIA committed:
refactor: 모달 리펙토링

- 모달 컴포넌트의 사용 용도에 맞는 위치로 경로 수정
- 새로운 API구조 적용
- Button component에서 전달받는 customStyle을 className으로 수정

JIPGERIA committed:
refactor: ModalInput & Modals 리펙토링

- ModalInput 컴포넌트에 Label을 추가하여 Label, Input을 한 번에 관리할 수 있도록 수정
